src/Intl/Idn/Punycode/CodePointString.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Polyfill\Intl\Idn\Punycode;
+
+use ArrayIterator;
+use IteratorAggregate;
+
+/**
+ * Copyright (c) 2020 Trevor Rowbotham
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Originally forked from
+ * https://github.com/TRowbotham/punycode/tree/1.0.0
+ *
+ * @author Trevor Rowbotham <trevor.rowbotham@pm.me>
+ */
+class CodePointString implements IteratorAggregate
+{
+    /**
+     * @var array<int, int>
+     */
+    protected $codePoints;
+
+    /**
+     * @param string $input
+     */
+    public function __construct($input)
+    {
+        $this->codePoints = $this->utf8Decode($input);
+    }
+
+    /**
+     * @return \ArrayIterator<int, int>
+     */
+    public function getIterator()
+    {
+        return new ArrayIterator($this->codePoints);
+    }
+
+    /**
+     * Takes a UTF-8 encoded string and converts it into a series of integer code points. Any
+     * invalid byte sequences will be replaced by a U+FFFD replacement code point.
```

Comment:
Unless I missed something, this "replacement" behavior is opening to security issues and is not compliant with the spec. See https://encoding.spec.whatwg.org/#security-background for an example issue.
Can't we use `array_map('mb_ord', preg_split('//u'...))` instead btw?

Reply:
The "replacement" behavior is correct. The decoder is meant to be run in the context of the [run](https://encoding.spec.whatwg.org/#concept-encoding-run) steps, which calls the [processing](https://encoding.spec.whatwg.org/#concept-encoding-process) steps. According to process Step 6, any time the decoder emits an error it would replace the code point with U+FFFD, so I just made the decoder emit a U+FFFD instead of an error. The other option would be to fail hard if we are given an invalid UTF-8 sequence, which would match the "fatal" condition in the run/process steps, but the default behavior is "replacement" for decoders.

> Can't we use `array_map('mb_ord', preg_split('//u'...))` instead btw?

I'm not sure off hand how `preg_split` handles invalid byte sequences, but I'll give it a shot.

Reply:
Just did a quick check and `preg_split('//u'...)` returns `false` when given an invalid UTF-8 string. Tested with `"\xC3\x7F\u{D800}"`.

Doing the following should have the same replacement behavior as `CodePointString::utf8Decode()` if you would prefer to do that:
```php
$input = '...';
mb_substitute_character(0xFFFD);
$input = mb_scrub($input, 'utf-8'); // doing mb_convert_encoding($input, 'utf-8', 'utf-8') should also work the same
$codePoints = array_map('mb_ord', preg_split('//u'...));
```

Reply:
how do the native functions behave? that should be the deciding criteria on the topic.
(btw, unrelated, did you forget to push the new mappings?)

Reply:
> how do the native functions behave? that should be the deciding criteria on the topic.

well, that is a somewhat complicated answer. The behavior depends on which ICU version is being used, as the behavior changed starting with ICU version 60.1. This happened in https://unicode-org.atlassian.net/browse/ICU-13311. Regardless of which ICU version is in use, there is a "replacement" behavior.

The old behavior was to replace each surrogate code point with a single U+FFFD replacement character. The new behavior, which adopted the behavior defined in the W3C/WHATWG spec (and also what `mb_scrub` does), is to replace each byte of each surrogate code point with a U+FFFD replacement character. The behavior of replacing invalid bytes that do not represent a surrogate code point remains unchanged, which is that each invalid byte gets replaced with a U+FFFD replacement character.

```php
$source = "\xC3a\u{D800}";
idn_to_utf8($source, ..., $info);

// In ICU versions < 60.1
echo $info['result'] ;  // prints "\u{FFFD}a\u{FFFD}"

// In ICU versions >= 60.1
echo $info['result']; // prints "\u{FFFD}a\u{FFFD}\u{FFFD}\u{FFFD}"
```

To achieve the old behavior, you would need to add an additional preprocessing step to replace all surrogates first, such as  `$input = preg_replace('/\xED[\xA0-\xBF][\x80-\xBF]/', "\u{FFFD}", $input);`, then run `mb_scrub` or whatever we choose to do the rest of the invalid byte replacements.

I went with the newer behavior since that seemed to reflect what `mbstring` did, but the older behavior technically isn't wrong either.  I tested both behaviors and the test suite doesn't seem to care. Let me know if you have a preference on which behavior to use.

It would appear that `mbstring`'s behavior has changed slightly over time, but mostly seems to adhere to the newer behavior. https://3v4l.org/ADc2e

> (btw, unrelated, did you forget to push the new mappings?)

No, just haven't gotten around to it yet :)

Reply:
Ok, thanks. Newest behavior is enough.